### PR TITLE
`close_to` matcher

### PR DESCRIPTION
**Pull request by danodonovan**
Implements the `close_to` matcher, in a similar way to pyhamcrest. Issue #1

### CODE REVIEW

File: precisely/comparison_matchers.py [hunk 1]
```diff
@@ -45,3 +46,22 @@ def match(self, actual):
...
+    return IsCloseToMatcher(value, delta)
+
+
+class IsCloseToMatcher(object):
+    def __init__(self, value, delta):
+        self._value = value
+        self._delta = delta
+
+    def match(self, actual):
+        if fabs(actual - self._value) <= self._delta:
```
Comment (mwilliamson) [inline precisely/comparison_matchers.py]: Is there a reason to use `fabs` rather than `abs`?

Come to think of it, avoiding `abs` might be useful to avoid unnecessarily not supporting other data types. For instance, you could imagine wanting to do a similar operation using datetimes.

Perhaps the check could be `value - delta <= actual <= value + delta`?

File: tests/numeric_comparison_tests.py [hunk 2]
```diff
@@ -58,3 +58,17 @@ def less_than_or_equal_to_matches_when_actual_is_less_than_or_equal_to_value():
 def less_than_or_equal_to_description_describes_value():
     matcher = less_than_or_equal_to(42)
     assert_equal("less than or equal to 42", matcher.describe())
+
+
+@istest
+def close_to_matches_when_actual_is_close_to_value_plus_delta():
+    matcher = close_to(42, 1)
+    assert_equal(matched(), matcher.match(43))
+    assert_equal(matched(), matcher.match(41))
```
Comment (mwilliamson) [inline tests/numeric_comparison_tests.py]: I think these two positive cases only test the boundaries. Could we also test a value not on the boundary?

File: tests/numeric_comparison_tests.py [hunk 3]
```diff
@@ -58,3 +58,17 @@ def less_than_or_equal_to_matches_when_actual_is_less_than_or_equal_to_value():
...
+    assert_equal(unmatched("was 40"), matcher.match(40))
```
Comment (mwilliamson) [inline tests/numeric_comparison_tests.py]: Is it worth including a distance in the error message, or is it just adding noise? e.g. `was 40 (2 away from 42)`. I'm not sure that's exactly the right format.

File: precisely/comparison_matchers.py (hunk 1 above)
Comment (danodonovan) [reply precisely/comparison_matchers.py]: Hamcrest use `fabs` - I assume so that the matcher does *only* support numeric data types (it's in `hamcrest.library.number`). I agree it's better to remove that limitation.

Comment (danodonovan) [reply precisely/comparison_matchers.py]: Further - I'll move the tests out of `numeric_comparison_tests.py` (into `comparison_tests.py` ?) and add a `datetime` test.

File: tests/numeric_comparison_tests.py (hunk 3 above)
Comment (danodonovan) [reply tests/numeric_comparison_tests.py]: Hamcrest have 
    
    <40> differed by <2>

Experimenting with 

    was 40 (differs from 42 by more than 1)

but it is noisier, and not very pleasant for the datetimes

**Comment by mwilliamson** [comment]
Hmm, I think including the actual difference is useful, rather than
"more than <delta>".

I'm on the fence about whether to include the expected value. Either is
fine by me, and we can always change it later.

File: precisely/comparison_matchers.py [hunk 4]
```diff
@@ -45,3 +45,22 @@ def match(self, actual):
...
+    return IsCloseToMatcher(value, delta)
+
+
+class IsCloseToMatcher(object):
+    def __init__(self, value, delta):
+        self._value = value
+        self._delta = delta
+
+    def match(self, actual):
+        if actual - self._delta <= self._value <= actual + self._delta:
```
Comment (mwilliamson) [inline precisely/comparison_matchers.py]: Hmm, `value - delta <= actual <= value + delta` seems clearer to me (in that that's how I'd explain what's going on to somebody if I was talking to them), but happy to persuaded that as-is is clearer (or that my suggestion is just plain wrong).

File: precisely/comparison_matchers.py [hunk 5]
```diff
@@ -45,3 +45,22 @@ def match(self, actual):
...
+class IsCloseToMatcher(object):
+    def __init__(self, value, delta):
+        self._value = value
+        self._delta = delta
+
+    def match(self, actual):
+        if self._value - self._delta <= actual <= self._value + self._delta:
+            return matched()
+        else:
+            return unmatched("was {0!r} ({1!r} away from {2!r})".format(actual, abs(self._value - actual), self._value))
```
Comment (mwilliamson) [inline precisely/comparison_matchers.py]: Thinking about this a bit more, we should be consistent in how we determine the gap. If we're requiring support for `abs()` (which I don't think we were in previous commits), then we should use the same value throughout i.e.

```python
difference = abs(self._value - actual)
if difference > self._delta:
    return unmatched("was {0!r} ({1!r} away from {2!r})".format(actual, difference, self._value))
else:
    return matched()
```

Apologies for the back and forth! Other than this, I think this is good to go.

Comment (danodonovan) [reply precisely/comparison_matchers.py]: The `abs` only went in to avoid negative distances - will update the rest of the code.
